### ISSUE TITLE
sbt plugin

### ISSUE THREAD

**Issue by vanDonselaar** [olafurpg/scala-db-codegen#9]
Nice to have: a sbt plugin for scala-db-codegen

Things to consider before coding:
- Create separate project with this project as dependency
- How to incorporate the task in the workflow: managed vs unmanaged sources

**Comment by olafurpg** [comment]
I agree that it could be easier to integrate the code generation into a project. However, I am not convinced that an sbt plugin is the way to go because:

1. iterating on the code generation setup is slow because you need to reload the entire build on every change.
2. I have bad experience maintaining the scalafmt sbt plugin. The SBT plugin system has several fundamental flaws IMO which cause a long tail of issues that you have to deal with as a maintainer
  - all plugins are on the same classpath, which causes sporadic issues
  - there is no way to require a minimum required sbt version
  - plugins must compile against scala 2.10, which is problematic if your library is 2.11 only (like scala-db-codegen)
  - many users don't understand how sbt works so you often end up having to explain sbt basics

Regarding managed vs unmanaged sources, I personally think you should check generated sources into version control. Why? Because it means you need a database setup to compile your project. I think `sbt compile` should work regardless if you have a db setup or not. Also, the generated code serves a decent documentation when browsing the repo online like on Github.

Instead of an sbt plugin, one direction we could go is to recommend using the API directly. Users could create a `db-codegen` project with a single dependency on scala-db-codegen. Then you write a tiny `main` function in that project that sets up all the configurations and runs scala-db-codegen.

WDYT?
